### PR TITLE
gnupg 2.2.0

### PR DESCRIPTION
**Pull request by fxcoudert**
Release notes: https://lists.gnupg.org/pipermail/gnupg-announce/2017q3/000413.html

poke @DomT4

### CODE REVIEW

File: Formula/gnupg.rb [hunk 1]
```diff
@@ -1,9 +1,9 @@
 class Gnupg < Formula
   desc "GNU Pretty Good Privacy (PGP) package"
   homepage "https://www.gnupg.org/"
-  url "https://gnupg.org/ftp/gcrypt/gnupg/gnupg-2.1.23.tar.bz2"
-  mirror "https://www.mirrorservice.org/sites/ftp.gnupg.org/gcrypt/gnupg/gnupg-2.1.23.tar.bz2"
-  sha256 "a94476391595e9351f219188767a9d6ea128e83be5ed3226a7890f49aa2d0d77"
+  url "https://www.gnupg.org/ftp/gcrypt/gnupg/gnupg-2.2.0.tar.bz2"
```
Comment (ilovezfs) [inline Formula/gnupg.rb]: World Wide Web!

**Comment by DomT4** [comment]
Aha, you literally beat me to it by about a minute 😄. There's breakage though, and the alias needs updating.

**Comment by DomT4** [comment]
To get it to build I had to do this, for some bizarre reason:

```diff
diff --git a/Formula/gnupg.rb b/Formula/gnupg.rb
index d8c08f096b..65f1c42746 100644
--- a/Formula/gnupg.rb
+++ b/Formula/gnupg.rb
@@ -49,6 +49,13 @@ class Gnupg < Formula
     args << "--with-readline=#{Formula["readline"].opt_prefix}" if build.with? "readline"
     args << "--enable-large-secmem" if build.with? "large-secmem"
 
+    # http.c:94:11: fatal error: 'gnutls/gnutls.h' file not found
+    # For some reason despite already being declared as a dependency
+    # the path to GnuTLS is getting dropped from the flags later in
+    # the compile process & this breaks the build.
+    # TODO: Try to work out why this happens & what to report upstream.
+    ENV.prepend "CPPFLAGS", "-I#{Formula["gnutls"].opt_include}"
+
     system "./configure", *args
     system "make"
     system "make", "check"
```

**Comment by ilovezfs** [comment]
>There's breakage though

Curious minds want to know

**Comment by DomT4** [comment]
See directly above 😉. The actual build error is:

```
clang -std=gnu99 -I/usr/local/opt/libgcrypt/include -I/usr/local/opt/libgpg-error/include -I/usr/local/opt/libksba/include -I/usr/local/opt/libgpg-error/include -I/usr/local/opt/libassuan/include -I/usr/local/opt/libgpg-error/include -I/usr/local/opt/libgpg-error/include -I/usr/local/Cellar/npth/1.5/include  -I/usr/local/Cellar/libtasn1/4.12/include -I/usr/local/Cellar/nettle/3.3/include -I/usr/local/Cellar/p11-kit/0.23.8/include/p11-kit-1 -I/usr/local/Cellar/gnutls/3.5.15/include -Wall -Wno-pointer-sign -Wpointer-arith  -g -O2   -o dirmngr-client dirmngr-client.o ../common/libcommon.a -L/usr/local/opt/libassuan/lib -lassuan -L/usr/local/opt/libgpg-error/lib -lgpg-error -L/usr/local/opt/libgpg-error/lib -lgpg-error -L/usr/local/opt/libgcrypt/lib -lgcrypt -L/usr/local/opt/libgpg-error/lib -lgpg-error  -lintl -Wl,-framework -Wl,CoreFoundation -liconv
http.c:94:11: fatal error: 'gnutls/gnutls.h' file not found
# include <gnutls/gnutls.h>
          ^~~~~~~~~~~~~~~~~
clang -I/usr/local/opt/libgpg-error/include -I/usr/local/opt/libgcrypt/include -I/usr/local/opt/libgpg-error/include -Wall -Wno-pointer-sign -Wpointer-arith  -g -O2   -o dirmngr_ldap dirmngr_ldap-dirmngr_ldap.o  ../common/libcommon.a -L/usr/local/opt/libgpg-error/lib -lgpg-error -L/usr/local/opt/libgcrypt/lib -lgcrypt -L/usr/local/opt/libgpg-error/lib -lgpg-error -lldap -llber -lintl -Wl,-framework -Wl,CoreFoundation -liconv
1 error generated.
make[2]: *** [t_ldap_parse_uri-http.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

**Comment by ilovezfs** [comment]
That is hideous. I'd like to see that fixed before this is merged.

**Comment by DomT4** [comment]
I'm slightly confused as to how it is even happening, because the `-I` flag to GnuTLS' headers is passed as part of that command, but apparently getting ignored by either the compiler or the `dirmngr` build process.

**Comment by ilovezfs** [comment]
It could be a Makefile bug.

**Comment by DomT4** [comment]
Of course, I am running macOS 10.13/Xcode 9, so that's another potential variable.

**Comment by DomT4** [comment]
Although I haven't seen anything else like that yet, and I have been building everything from source lately due to the lack of High Sierra bottles thus far. But maybe I've just been getting lucky 🤷‍♂️.

**Comment by fxcoudert** [comment]
Build is passing on CI, macOS 10.10/11/12.

**Comment by ilovezfs** [comment]
@DomT4 did you try just setting --with-libgcrypt-prefix=

**Comment by DomT4** [comment]
Huh. Maybe we've stumbled upon a subtle problem with Xcode 9.

> did you try just setting --with-libgcrypt-prefix=

No, but the issue is with finding GnuTLS, unless I'm missing your point here? I slept awfully overnight, so you may need to ELI5 😅.

**Comment by ilovezfs** [comment]
No I just cannot read.

**Comment by ilovezfs** [comment]
@fxcoudert do you mind checking if the build passes on your 10.13 box?

**Comment by fxcoudert** [comment]
@BrewTestBot test this please

**Comment by ilovezfs** [comment]
@DomT4 is your gnutls linked? I can reproduce it by unlink gnutls. So it's probably only working because the default prefix is /usr/local. I bet this build breaks on < 10.13 in a non-standard prefix.

**Comment by DomT4** [comment]
> is your gnutls linked?

Ooft. No, it isn't. I forgot I unlinked it chasing some issues with building Tunnelblick locally the other week.

**Comment by ilovezfs** [comment]
>I'm slightly confused as to how it is even happening, because the -I flag to GnuTLS' headers is passed as part of that command, but apparently getting ignored by either the compiler or the dirmngr build process.

I think it's because it's the compilation of http.c that is failing and you're looking at the wrong compilation line.

```
clang -DHAVE_CONFIG_H -I. -I..  -DLOCALEDIR=\"/usr/local/Cellar/gnupg/2.2.0/share/locale\" -DGNUPG_BINDIR="\"/usr/local/Cellar/gnupg/2.2.0/bin\"" -DGNUPG_LIBEXECDIR="\"/usr/local/Cellar/gnupg/2.2.0/libexec\"" -DGNUPG_LIBDIR="\"/usr/local/Cellar/gnupg/2.2.0/lib/gnupg\"" -DGNUPG_DATADIR="\"/usr/local/Cellar/gnupg/2.2.0/share/gnupg\"" -DGNUPG_SYSCONFDIR="\"/usr/local/etc/gnupg\"" -DGNUPG_LOCALSTATEDIR="\"/usr/local/Cellar/gnupg/2.2.0/var\""  -DGNUPG_DEFAULT_PINENTRY="\"/usr/local/opt/pinentry/bin/pinentry\""      -DWITHOUT_NPTH=1  -std=gnu99 -I/usr/local/opt/libgcrypt/include -I/usr/local/opt/libgpg-error/include -I/usr/local/opt/libassuan/include -I/usr/local/opt/libgpg-error/include -I/usr/local/opt/libgpg-error/include -Wall -Wno-pointer-sign -Wpointer-arith  -g -O2 -c -o t_ldap_parse_uri-http.o `test -f 'http.c' || echo './'`http.c
http.c:94:11: fatal error: 'gnutls/gnutls.h' file not found
# include <gnutls/gnutls.h>
          ^
1 error generated.
make[2]: *** [t_ldap_parse_uri-http.o] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```
No gnutls include path is passed.

**Comment by ilovezfs** [comment]
>BrewTestBot test this please

why are we re-running it?

**Comment by ilovezfs** [comment]
```
t_ldap_parse_uri_SOURCES = \
        t-ldap-parse-uri.c ldap-parse-uri.c ldap-parse-uri.h \
        http.c http-common.c dns-stuff.c \
        $(ldap_url) $(t_common_src)
t_ldap_parse_uri_CFLAGS = -DWITHOUT_NPTH=1  $(USE_C99_CFLAGS) \
                          $(LIBGCRYPT_CFLAGS) \
                          $(LIBASSUAN_CFLAGS) $(GPG_ERROR_CFLAGS)
```
No LIBGNUTLS_CFLAGS.

**Comment by DomT4** [comment]
I wonder if this is a new bug or just one we somehow haven't noticed before.

**Comment by ilovezfs** [comment]
```
diff --git a/Formula/gnupg.rb b/Formula/gnupg.rb
index c205db539d..5535787f50 100644
--- a/Formula/gnupg.rb
+++ b/Formula/gnupg.rb
@@ -18,6 +18,9 @@ class Gnupg < Formula
 
   deprecated_option "without-libusb-compat" => "without-libusb"
 
+  depends_on "autoconf" => :build
+  depends_on "automake" => :build
+  depends_on "libtool" => :build
   depends_on "pkg-config" => :build
   depends_on "sqlite" => :build if MacOS.version == :mavericks
   depends_on "npth"
@@ -33,6 +36,8 @@ class Gnupg < Formula
   depends_on "readline" => :optional
   depends_on "encfs" => :optional
 
+  patch :DATA
+
   def install
     args = %W[
       --disable-dependency-tracking
@@ -49,6 +54,7 @@ class Gnupg < Formula
     args << "--with-readline=#{Formula["readline"].opt_prefix}" if build.with? "readline"
     args << "--enable-large-secmem" if build.with? "large-secmem"
 
+    system "autoreconf", "-fiv"
     system "./configure", *args
     system "make"
     system "make", "check"
@@ -96,3 +102,18 @@ class Gnupg < Formula
     end
   end
 end
+
+__END__
+diff --git a/dirmngr/Makefile.am b/dirmngr/Makefile.am
+index b404165..35cc424 100644
+--- a/dirmngr/Makefile.am
++++ b/dirmngr/Makefile.am
+@@ -152,7 +152,7 @@ t_ldap_parse_uri_SOURCES = \
+         http.c http-common.c dns-stuff.c \
+         $(ldap_url) $(t_common_src)
+ t_ldap_parse_uri_CFLAGS = -DWITHOUT_NPTH=1  $(USE_C99_CFLAGS) \
+-			  $(LIBGCRYPT_CFLAGS) \
++			  $(LIBGCRYPT_CFLAGS) $(LIBGNUTLS_CFLAGS) \
+                           $(LIBASSUAN_CFLAGS) $(GPG_ERROR_CFLAGS)
+ t_ldap_parse_uri_LDADD = $(ldaplibs) $(t_common_ldadd) $(DNSLIBS)
+ 
```
fixes it

**Comment by DomT4** [comment]
Is it worth using my hack instead to avoid introducing the autotool deps for this release, since both seem to work? Obviously not suggesting we don't report your fix upstream & have them carry it in future.

**Comment by ilovezfs** [comment]
I think we can just leave it broken as long as it gets reported.

**Comment by fxcoudert** [comment]
One failure, `duplicity` test fails on all macOS versions:

```
Error: No GPG present to test against!
Error: duplicity: failed
```

**Comment by ilovezfs** [comment]
No it's not new. 2.1 fails the same way.

**Comment by DomT4** [comment]
Oh I know why that's failing now. Give me a bit, need a drink, then I'll whack a PR your way.

**Comment by ilovezfs** [comment]
gpg_version == Version.create("2.1") ||
        gpg_version == Version.create("2.0")

**Comment by DomT4** [comment]
Yup.

**Comment by ilovezfs** [comment]
after fixing that I get an even more beautiful failure
```
iMac-TMP:Homebrew joe$ brew test duplicity
Testing duplicity
gpg: directory '/tmp/duplicity-test-20170828-99215-bxsl1i/.gnupg' created
gpg: keybox '/tmp/duplicity-test-20170828-99215-bxsl1i/.gnupg/pubring.kbx' created
gpg: /tmp/duplicity-test-20170828-99215-bxsl1i/.gnupg/trustdb.gpg: trustdb created
gpg: key 574DFC639641F6C6 marked as ultimately trusted
gpg: directory '/tmp/duplicity-test-20170828-99215-bxsl1i/.gnupg/openpgp-revocs.d' created
gpg: revocation certificate stored as '/tmp/duplicity-test-20170828-99215-bxsl1i/.gnupg/openpgp-revocs.d/58D4975B5995EB26546BB515574DFC639641F6C6.rev'
==> ./command.sh
Last 15 lines from /Users/joe/Library/Logs/Homebrew/duplicity/test.01.command.sh:
2017-08-28 08:40:33 -0700

./command.sh

spawn /usr/local/Cellar/duplicity/0.7.13.1/bin/duplicity /tmp/duplicity-test-20170828-99215-bxsl1i file://test
Max open files of 256 is too low, should be >= 1024.
Use 'ulimit -n 1024' or higher to correct.

expect: spawn id exp7 not open
    while executing
"expect -exact "Last full backup date: none""
    (file "./command.sh" line 5)
Error: duplicity: failed
Failed executing: ./command.sh 
/usr/local/Homebrew/Library/Homebrew/formula.rb:1804:in `block in system'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1742:in `open'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1742:in `system'
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/duplicity.rb:409:in `block in <class:Duplicity>'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1639:in `block (2 levels) in run_test'
/usr/local/Homebrew/Library/Homebrew/formula.rb:834:in `with_logging'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1638:in `block in run_test'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:14:in `block in mktemp'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:74:in `block in run'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:74:in `chdir'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:74:in `run'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:13:in `mktemp'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1632:in `run_test'
/usr/local/Homebrew/Library/Homebrew/test.rb:28:in `block in <main>'
/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.0.0-p648/lib/ruby/2.0.0/timeout.rb:66:in `timeout'
/usr/local/Homebrew/Library/Homebrew/test.rb:27:in `<main>'
iMac-TMP:Homebrew joe$ 
```

**Comment by ilovezfs** [comment]
```
iMac-TMP:homebrew-core joe$ git diff -- Formula/duplicity.rb 
diff --git a/Formula/duplicity.rb b/Formula/duplicity.rb
index 8a13b23..9f2cd18 100644
--- a/Formula/duplicity.rb
+++ b/Formula/duplicity.rb
@@ -406,7 +406,7 @@ class Duplicity < Formula
       expect eof
     EOS
     chmod 0755, testpath/"command.sh"
-    system "./command.sh"
+    system "ulimit -n 1024; ./command.sh"
     assert_match "duplicity-full-signatures", Dir["test/*"].to_s
 
     # Ensure requests[security] is activated
```
works.

But ... ew.

**Comment by ilovezfs** [comment]
So do we want 2.2 || 2.0, or 2.2 || 2.1 || 2.0

**Comment by DomT4** [comment]
Thoughts welcome @ https://github.com/Homebrew/brew/pull/3098.

**Comment by ilovezfs** [comment]
pulling so there's no rush on the other stuff
